### ISSUE TITLE
CentralProcesor getProcessorID is different from “sudo dmidecode -t 4 | grep 'ID'”

### ISSUE THREAD

**Issue by ustcck** [oshi/oshi#2392]
JDK version： OpenJDK_17.0.7_7
oshi version： 6.4.0
OS：Ubuntu 22.04.2

// 078BFBFF00500654
![image](https://github.com/oshi/oshi/assets/16115192/c6cf1fb7-6308-4713-a984-3e3c27896abe)

> // 078BFBFF00050654
> caokai@caokai:~$ sudo dmidecode -t processor | grep 'ID' | awk -F ':' '{print $2}' | head -n 1
>  54 06 05 00 FF FB 8B 07

**Comment by dbwiddis** [comment]
This is the same CPUID with the endianness reversed

07 8B FB FF 00 05 06 54

54 06 05 00 FF FB 8B 07

[This is documented in the javadoc.](https://www.oshi.ooo/oshi-core-java11/apidocs/com.github.oshi/oshi/hardware/CentralProcessor.ProcessorIdentifier.html#getProcessorID())
> NOTE: The order of returned bytes is platform and software dependent. Values may be in either Big Endian or Little Endian order.

**Comment by ustcck** [comment]
> This is the same CPUID with the endianness reversed
> 
> 07 8B FB FF 00 05 06 54
> 
> 54 06 05 00 FF FB 8B 07
> 
> [This is documented in the javadoc.](https://www.oshi.ooo/oshi-core-java11/apidocs/com.github.oshi/oshi/hardware/CentralProcessor.ProcessorIdentifier.html#getProcessorID())
> 
> > NOTE: The order of returned bytes is platform and software dependent. Values may be in either Big Endian or Little Endian order.

I know，078BFBFF00**05**0654 && 078BFBFF00**50**0654

**Comment by ustcck** [comment]
I know，078BFBFF00_05_0654 && 078BFBFF00_50_0654 ，two results

**Comment by ustcck** [comment]
Look closely, two results

**Comment by dbwiddis** [comment]
> Look closely, two results

Fair enough but you have listed both:
```
// 078BFBFF00050654
```
and
```
// 078BFBFF00500654
```
Which is OSHI output?

To be clear:
1. OSHI reports CPUID from OS sources which may differ in endianness and whitespace
2. CPUID is an Intel-only thing that we report as a best-effort
3. On Linux, without sudo permission, we have to recreate the CPUID from known values of family, model, and stepping (which is essentially all the CPUID is) and there is only one bit that we have to guess at.

TLDR: please be clear on exactly what OSHI methods are producing wrong output, as it's not clear from what you've indicated.

**Comment by ustcck** [comment]
Sorry

078BFBFF00500654 is OSHI output，“createProcessorID(stepping, model, family, flags)”
078BFBFF00050654 is Dmidecode output，“sudo dmidecode -t processor | grep 'ID' | awk -F ':' '{print $2}' | head -n 1”

**Comment by dbwiddis** [comment]
OK, thanks for clarifying.  I do see that same output (ignoring the flags for now):
```java
System.out.println(createProcessorID("4", "85", "6", new String[0]));
```
> 0000000000500654

So the 5 (0101) is in bits  20-23 instead of 16-19 according to `dmidecode` output.

OSHI's bits shift left to 16...
```java
        // 19:16,7:4 – Model
        processorIdBytes |= (modelL & 0x0f) << 4;
        processorIdBytes |= (modelL & 0xf0) << 16;
```
but they don't account for the facct that we're already dealing with the high 4 bits of the byte. That should only be a shift left of 12.   (There's a similar bug for family.)

Bit shifting is strange... thanks for the bug report!

**Comment by ustcck** [comment]
Many thanks for the fix!
